File: src/test/resources/checks/PerformanceArrayOperationsInLoopCheck.php
/*
 * Copyright (C) 2025 Rostislav Suleimanov
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

<?php

class ArrayOperationsTest
{
    public function nonCompliantArrayOperations(): void
    {
        $result = [];
        $items = $this->getItems();


        foreach ($items as $item) {
            $result = array_merge($result, $item->getValues()); // Noncompliant {{array_merge() is used in a loop and is a resource-intensive operation.}}
        }

        $datasets = $this->getDatasets();
        $combinedData = [];


        for ($i = 0; $i < count($datasets); $i++) {
            $combinedData = array_merge_recursive($combinedData, $datasets[$i]); // Noncompliant {{array_merge_recursive() is used in a loop and is a resource-intensive operation.}}
        }

        $counter = 0;
        $masterArray = ['a' => 1, 'b' => 2, 'c' => 3];
        $differences = [];


        while ($counter < 10) {
            $testArray = $this->getRandomArray();
            $differences[] = array_diff($masterArray, $testArray); // Noncompliant {{array_diff() is used in a loop and is a resource-intensive operation.}}
            $counter++;
        }

        $index = 0;
        $comparisonArrays = $this->getComparisonArrays();

        do {
            $commonElements = array_intersect($masterArray, $comparisonArrays[$index]);  // Noncompliant {{array_intersect() is used in a loop and is a resource-intensive operation.}}
            $this->processCommonElements($commonElements);
            $index++;
        } while ($index < count($comparisonArrays));


        foreach ($this->getConfigurations() as $config) {
            $defaultConfig = $this->getDefaultConfig();
            $activeConfig = array_replace($defaultConfig, $config); // Noncompliant {{array_replace() is used in a loop and is a resource-intensive operation.}}
            $this->applyConfiguration($activeConfig);
        }
    }

    public function compliantArrayOperations(): void
    {
        $collections = [];
        $items = $this->getItems();
        foreach ($items as $item) {
            $collections[] = $item->getValues();
        }
        $result = array_merge(...$collections); // Compliant: Collect all items first, then merge once outside the loop

        $datasets = $this->getDatasets();
        $combinedData = [];
        for ($i = 0; $i < count($datasets); $i++) {
            foreach ($datasets[$i] as $key => $value) {
                $combinedData[$key] = $value; // Compliant: Direct array addition within the loop
            }
        }


        $configs = $this->getConfigurations();
        $defaultConfig = $this->getDefaultConfig();
        foreach ($configs as $config) {
            $activeConfig = $defaultConfig + $config; // Compliant: Using array union operator instead of array_merge
            $this->applyConfiguration($activeConfig);
        }


        $largeDataset = $this->getLargeDataset();
        $chunkSize = 100;
        $chunks = array_chunk($largeDataset, $chunkSize);

        foreach ($chunks as $chunk) {
            $processedChunk = $this->processDataChunk($chunk); // Compliant: Process in chunks to avoid loop with expensive operations
            $this->saveResults($processedChunk);
        }
    }

    public function mixedArrayOperationsExample(): void
    {
        $items = $this->getItems();

        $result1 = [];

        foreach ($items as $item) {
            if ($item->isSpecial()) {
                $result1 = array_merge($result1, $item->getSpecialValues()); // Noncompliant {{array_merge() is used in a loop and is a resource-intensive operation.}}
            }
        }

        $result2 = [];
        $specialValues = [];
        foreach ($items as $item) {
            if ($item->isImportant()) {
                $specialValues[] = $item->getImportantValues();
            }
        }
        if (!empty($specialValues)) {
            $result2 = array_merge(...$specialValues); //Compliant: Single merge operation outside the loop
        }


        $masterConfig = $this->getMasterConfig();
        foreach ($this->getUserConfigs() as $userConfig) {
            $differences = array_diff_assoc($masterConfig, $userConfig); // Noncompliant {{array_diff_assoc() is used in a loop and is a resource-intensive operation.}}
            $this->reportConfigDifferences($differences);
        }
    }

    private function getItems(): array { return []; }
    private function getDatasets(): array { return []; }
    private function getRandomArray(): array { return []; }
    private function getComparisonArrays(): array { return []; }
    private function getConfigurations(): array { return []; }
    private function getDefaultConfig(): array { return []; }
    private function getLargeDataset(): array { return []; }
    private function getMasterConfig(): array { return []; }
    private function getUserConfigs(): array { return []; }

    private function processCommonElements(array $elements): void {}
    private function applyConfiguration(array $config): void {}
    private function processDataChunk(array $chunk): array { return []; }
    private function saveResults(array $results): void {}
    private function reportConfigDifferences(array $diff): void {}
}

File: src/main/java/org/rostilos/sonarrules/php/checks/PerformanceArrayOperationsInLoopCheck.java
/*
 * Copyright (C) 2025 Rostislav Suleimanov
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.rostilos.sonarrules.php.checks;

import org.sonar.check.Rule;
import org.sonar.check.Priority;
import org.sonar.plugins.php.api.tree.Tree;
import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;


@Rule(
        key = PerformanceArrayOperationsInLoopCheck.KEY,
        name = "Resource-intensive array operations should not be used in loops",
        description = "Resource-intensive array operations like array_merge() in loops can cause significant performance issues.",
        priority = Priority.MAJOR,
        tags = {"performance", "magento2"}
)
public class PerformanceArrayOperationsInLoopCheck extends PHPVisitorCheck {

    public static final String KEY = "PerformanceArrayOperationsInLoop";
    private static final Set<String> INEFFICIENT_ARRAY_FUNCTIONS = new HashSet<>(Arrays.asList(
            "array_merge",
            "array_merge_recursive",
            "array_replace",
            "array_replace_recursive",
            "array_diff",
            "array_diff_assoc",
            "array_diff_key",
            "array_diff_uassoc",
            "array_diff_ukey",
            "array_intersect",
            "array_intersect_assoc",
            "array_intersect_key",
            "array_intersect_uassoc",
            "array_intersect_ukey",
            "array_udiff",
            "array_udiff_assoc",
            "array_udiff_uassoc",
            "array_uintersect",
            "array_uintersect_assoc",
            "array_uintersect_uassoc"
    ));

    private static final String MESSAGE_TEMPLATE = "%s() is used in a loop and is a resource-intensive operation.";

    @Override
    public void visitFunctionCall(FunctionCallTree tree) {
        String functionName = tree.callee().toString();

        if (INEFFICIENT_ARRAY_FUNCTIONS.contains(functionName) && isInsideLoop(tree)) {
            context().newIssue(this, tree, String.format(MESSAGE_TEMPLATE, functionName));
        }

        super.visitFunctionCall(tree);
    }

    private boolean isInsideLoop(Tree tree) {
        Tree parent = tree.getParent();

        while (parent != null) {
            if (parent.is(Tree.Kind.FOR_STATEMENT) ||
                    parent.is(Tree.Kind.FOREACH_STATEMENT) ||
                    parent.is(Tree.Kind.WHILE_STATEMENT) ||
                    parent.is(Tree.Kind.DO_WHILE_STATEMENT)) {
                return true;
            }
            parent = parent.getParent();
        }

        return false;
    }
}
